File: RCTTest/ios/RCTTest/Common/UIColor+RGBA.m
//
//  UIColor+RGBA.m
//  RCTTest
//
//  Created by 童万华 on 17/8/5.
//  Copyright © 2017年 童万华. All rights reserved.
//

#import "UIColor+RGBA.h"

@implementation UIColor (RGBA)

+(NSArray *)colorForRGBAWithConvertColor:(UIColor *)color {
  CGFloat numberOfComponents = CGColorGetNumberOfComponents(color.CGColor);
    NSArray *rgbaComponents ;
    if (numberOfComponents == 4) {
        const CGFloat *components = CGColorGetComponents(color.CGColor);
        rgbaComponents = [NSArray arrayWithObjects:@(components[0]),@(components[1]),@(components[2]),@(components[3]), nil];
    }else if (numberOfComponents == 2) {
        if (CGColorEqualToColor(color.CGColor, [UIColor clearColor].CGColor)) {
            rgbaComponents = [NSArray arrayWithObjects:@(0.0),@(0.0),@(0.0),@(0),nil];
        }
        if (CGColorEqualToColor(color.CGColor, [UIColor blackColor].CGColor)) {
            rgbaComponents = [NSArray arrayWithObjects:@(0),@(0), @(0), @(1),nil];
        }
        if (CGColorEqualToColor(color.CGColor, [UIColor whiteColor].CGColor)) {
            rgbaComponents = [NSArray arrayWithObjects:@(1),@(1),@(1),@(1),nil];
        }
        if (CGColorEqualToColor(color.CGColor, [UIColor grayColor].CGColor)) {
            rgbaComponents = [NSArray arrayWithObjects:@(0.5),@(0.5),@(0.5),@(1),nil];
        }
        if (CGColorEqualToColor(color.CGColor, [UIColor lightGrayColor].CGColor)) {
            rgbaComponents = [NSArray arrayWithObjects:@(0.667),@(0.667),@(0.667),@(1),nil];
        }
        if (CGColorEqualToColor(color.CGColor, [UIColor darkGrayColor].CGColor)) {
            rgbaComponents = [NSArray arrayWithObjects:@(0.333),@(0.333),@(0.333),@(1),nil];
        }

    }
    return rgbaComponents;
}
+(NSArray *)colorForGradientRGBAWith:(NSArray *)normalRgbaComponents selectedRgbaComponents:(NSArray*)selectedRgbaComponents {
    NSArray *gradientRGBAArr;
    if (normalRgbaComponents && selectedRgbaComponents) {
        CGFloat deltaR = [normalRgbaComponents[0] floatValue] - [selectedRgbaComponents[0] floatValue];
        CGFloat deltaG = [normalRgbaComponents[1] floatValue] - [selectedRgbaComponents[1] floatValue];
        CGFloat deltaB = [normalRgbaComponents[2] floatValue] - [selectedRgbaComponents[2] floatValue];
        CGFloat deltaA = [normalRgbaComponents[3] floatValue] - [selectedRgbaComponents[3] floatValue];
        gradientRGBAArr = [NSArray arrayWithObjects:@(deltaR), @(deltaG), @(deltaB), @(deltaA), nil];
    }
    return gradientRGBAArr;
}

+(UIColor *)oldColorWithSelectedColorRGBA:(NSArray*)selectedColorRGBA deltaRGBA:(NSArray*)deltaRGBA scale:(CGFloat)scale
{
    return [UIColor colorWithRed:[selectedColorRGBA[0] floatValue] + [deltaRGBA[0] floatValue] * scale green:[selectedColorRGBA[1] floatValue] + [deltaRGBA[1] floatValue] * scale blue:[selectedColorRGBA[2] floatValue] + [deltaRGBA[2] floatValue] * scale alpha:[selectedColorRGBA[3] floatValue] + [deltaRGBA[3] floatValue] * scale];
}
+(UIColor *)newColorWithNormalColorRGBA:(NSArray*)normalColorRGBA deltaRGBA:(NSArray*)deltaRGBA scale:(CGFloat)scale
{
    return [UIColor colorWithRed:[normalColorRGBA[0] floatValue] - [deltaRGBA[0] floatValue] * scale green:[normalColorRGBA[1] floatValue] - [deltaRGBA[1] floatValue] * scale blue:[normalColorRGBA[2] floatValue] - [deltaRGBA[2] floatValue] * scale alpha:[normalColorRGBA[3] floatValue] - [deltaRGBA[3] floatValue] * scale];
}

@end

File: RCTTest/ios/RCTTest/Common/ScrollPageView(swift)/UIColor+RGBA.swift
//
//  UIColor+RGBA.swift
//  RCTTest
//
//  Created by 童万华 on 2017/8/22.
//  Copyright © 2017年 童万华. All rights reserved.
//

import Foundation
extension UIColor {
    //返回标准的RGBA色值数组
    func rgba() -> [CGFloat]? {
        guard let components = self.cgColor.components else {
            return nil
        }
        
        if components.count == 2 {
            if self == UIColor.clear {
                return [0.0,0.0,0.0,0.0]
            }else if self == UIColor.black {
                return [0.0,0.0,0.0,1.0]
            }else if self == UIColor.white {
                return [1.0,1.0,1.0,1.0]
            }else if self == UIColor.gray {
                return [0.5,0.5,0.5,1.0]
            }else if self == UIColor.darkGray {
                return [0.333,0.333,0.333,1.0]
            }else if self == UIColor.lightGray {
                return [0.667,0.667,0.667,1.0]
            }
        }
        return components
    }
    
   class func gradientRGBA(normalRgbaComponents:[CGFloat]?,selectedRgbaComponents:[CGFloat]?) -> [CGFloat]? {
        guard let normalRgbaComponents = normalRgbaComponents,
            let selectedRgbaComponents = selectedRgbaComponents  else {
                return nil
        }
        let deltaR = normalRgbaComponents[0] - selectedRgbaComponents[0]
        let deltaG = normalRgbaComponents[1] - selectedRgbaComponents[1]
        let deltaB = normalRgbaComponents[2] - selectedRgbaComponents[2]
        let deltaA = normalRgbaComponents[3] - selectedRgbaComponents[3]
        return [deltaR,deltaG,deltaB,deltaA]
    }
    
   class func transfromToNormalColor(selectedRgbaComponents:[CGFloat]?,deltaRGBA:[CGFloat]?,scale:CGFloat) -> UIColor? {
        guard let selectedRgbaComponents = selectedRgbaComponents,
            let deltaRGBA = deltaRGBA else {
                return nil
        }
        let normalR = selectedRgbaComponents[0] + deltaRGBA[0]*scale
        let normalG = selectedRgbaComponents[1] + deltaRGBA[1]*scale
        let normalB = selectedRgbaComponents[2] + deltaRGBA[2]*scale
        let normalA = selectedRgbaComponents[3] + deltaRGBA[3]*scale
        
        return UIColor(red: normalR, green: normalG, blue: normalB, alpha: normalA)
    }
   class func transfromToSelectedColor(normalRgbaComponents:[CGFloat]?,deltaRGBA:[CGFloat]?,scale:CGFloat) -> UIColor? {
        guard let normalRgbaComponents = normalRgbaComponents,
            let deltaRGBA = deltaRGBA else {
                return nil
        }
        let selectedR = normalRgbaComponents[0] - deltaRGBA[0]*scale
        let selectedG = normalRgbaComponents[1] - deltaRGBA[1]*scale
        let selectedB = normalRgbaComponents[2] - deltaRGBA[2]*scale
        let selectedA = normalRgbaComponents[3] - deltaRGBA[3]*scale
        
        return UIColor(red: selectedR, green: selectedG, blue: selectedB, alpha: selectedA)
    }
}
